pedsnetdcc/id_mapping_transform.py

```diff
@@ -0,0 +1,239 @@
+            raise ValueError('cannot map IDs for multi-column primary key'
+                             ' on table {0}'.format(table_name))
+
+        # Get primary key name and mapping table name, defined by convention.
+        pkey_name = list(table.primary_key.columns.keys())[0]
+        map_table_name = map_table_name_tmpl.format(table_name=table_name)
+
+        # Construct table object for mapping table.
+        map_table = sqlalchemy.Table(
+            map_table_name, metadata,
+            sqlalchemy.Column('dcc_id', sqlalchemy.Integer),
+            sqlalchemy.Column('site_id', sqlalchemy.Integer))
+
+        # Add a join to the mapping table.
+        if join:
+            join = join.join(map_table, table.c[pkey_name] ==
+                             map_table.c['site_id'])
+        else:
+            join = join or sqlalchemy.join(table, map_table,
+                                           table.c[pkey_name] ==
+                                           map_table.c['site_id'])
+
+        # Create a new select object, because we need to get rid of the
+        # primary key column.
+        select = select or sqlalchemy.select([table])
+        new_select = sqlalchemy.select()
+
+        # Get rid of the primary key column.
+        for c in select.inner_columns:
+            if c.name != pkey_name:
+                new_select.append_column(c)
+
+        # Add the mapping table dcc_id column as the primary key column.
+        new_select.append_column(map_table.c['dcc_id'].label(pkey_name))
+
+        # Put the new select object back in the original var for further use.
+        select = new_select
+
+        for fkey in table.foreign_key_constraints:
+
+            ref_table_name = fkey.referred_table.name
+
+            # Do not operate on foreign keys to vocabulary tables.
+            if ref_table_name in VOCAB_TABLES:
+                continue
+
+            # Get primary key name and mapping table name, defined by
+            # convention.
+            fkey_name = fkey.column_keys[0]
+            map_table_name = map_table_name_tmpl.\
+                format(table_name=ref_table_name)
+
+            # Construct table object for mapping table.
+            map_table = sqlalchemy.Table(
+                map_table_name, metadata,
+                sqlalchemy.Column('dcc_id', sqlalchemy.Integer),
+                sqlalchemy.Column('site_id', sqlalchemy.Integer))
+
+            # Make the join an outer join if the foreign key is nullable,
+            # otherwise rows without this foreign key will not be included in
+            # the final product.
+            isouter = table.c[fkey_name].nullable
+
+            # Add a join to the mapping table.
+            join = join.join(map_table, table.c[fkey_name] ==
+                             map_table.c['site_id'], isouter=isouter)
+
+            # Create a new select object, because we need to get rid of the
+            # primary key column.
```

Comment:
foreign instead of primary. Also, instead of 'get rid of', it would be more clear to mention 'we need to replace the foreign key column with one using the mapped value' or some such.